pulse2percept/api.py

```diff
@@ -0,0 +1,423 @@
+        else:
+            e_s = "Supported electrode types are 'epiretinal', 'subretinal'"
+            raise ValueError(e_s)
+
+        # Derive the effective current spread
+        if self.ofl_streaks:
+            ecs, _ = self.ofl.electrode_ecs(self.implant)
+        else:
+            _, ecs = self.ofl.electrode_ecs(self.implant)
+
+        # Calculate the max of every current spread map
+        lmax = np.zeros((2, ecs.shape[-1]))
+        if 'INL' in dolayers:
+            lmax[0, :] = ecs[:, :, 0, :].max(axis=(0, 1))
+        if 'NFL' in dolayers:
+            lmax[1, :] = ecs[:, :, 1, :].max(axis=(0, 1))
+
+        # `ecs_list` is a pixel by `n` list where `n` is the number of layers
+        # being simulated. Each value in `ecs_list` is the current contributed
+        # by each electrode for that spatial location
+        ecs_list = []
+        idx_list = []
+        for xx in range(self.ofl.gridx.shape[1]):
+            for yy in range(self.ofl.gridx.shape[0]):
+                # If any of the used current spread maps at [yy, xx] are above
+                # tolerance, we need to process that pixel
+                process_pixel = False
+                if 'INL' in dolayers:
+                    # For this pixel: Check if the ecs in any layer is large
+                    # enough compared to the max across pixels within the layer
+                    process_pixel |= np.any(ecs[yy, xx, 0, :] >=
+                                            tol * lmax[0, :])
+                if 'NFL' in dolayers:
+                    process_pixel |= np.any(ecs[yy, xx, 1, :] >=
+                                            tol * lmax[1, :])
+
+                if process_pixel:
+                    ecs_list.append(ecs[yy, xx])
+                    idx_list.append([yy, xx])
+
+        s_info = "tol=%.1f%%, %d/%d px selected" % (tol * 100, len(ecs_list),
+                                                    np.prod(ecs.shape[:2]))
+        logging.getLogger(__name__).info(s_info)
+
+        logging.getLogger(__name__).info("Starting computation...")
+        sr_list = p2p.utils.parfor(self.gcl.calc_per_pixel,
+                                   ecs_list, n_jobs=self.num_jobs,
+                                   engine=self.engine,
+                                   func_args=[pt_list, dolayers, self.dojit])
+        bm = np.zeros(self.ofl.gridx.shape +
+                      (sr_list[0].data.shape[-1], ))
+        idxer = tuple(np.array(idx_list)[:, i] for i in range(2))
+        bm[idxer] = [sr.data for sr in sr_list]
+        percept = p2p.utils.TimeSeries(sr_list[0].tsample, bm)
+
+        if t_percept != percept.tsample:
+            percept = percept.resample(t_percept)
+
+        logging.getLogger(__name__).info("Done.")
+
+        return percept
+
+    def set_ofl(self, sampling=100, axon_lambda=2, rot_deg=0,
+                x_range=None, y_range=None,
+                loadpath='../', save_data=True, streaks_enabled=True):

```

Comment:
This `loadpath` is a bit odd. Consider something more rational (e.g., `~/.pulse2percept` or `.`)